internal/checks/latency.go

```diff
@@ -0,0 +1,458 @@
+	defer cancel()
+
+	latencyBreaker := c.getLatencyCircuitBreaker(method)
+
+	// Make and record the request.
+	var duration time.Duration
+	duration, c.Err = c.client.Latency(ctx, method)
+	// TODO(polsar): The error must also pass the checks specified in the config
+	//  (i.e. match HTTP code, JSON RPC code, and error message).
+	//  Fixing this is not a priority since we're not currently using passive health checking.
+	isError := c.Err != nil
+	c.errorCircuitBreaker.RecordRequest(isError)
+	latencyBreaker.RecordLatency(duration)
+
+	if isError {
+		c.metricsContainer.LatencyCheckErrors.WithLabelValues(
+			c.upstreamConfig.ID,
+			c.upstreamConfig.HTTPURL,
+			metrics.HTTPRequest,
+		).Inc()
+	} else if duration >= latencyThreshold {
+		c.metricsContainer.LatencyCheckHighLatencies.WithLabelValues(
+			c.upstreamConfig.ID,
+			c.upstreamConfig.HTTPURL,
+			metrics.HTTPRequest,
+		).Inc()
+	}
+
+	c.metricsContainer.Latency.WithLabelValues(
+		c.upstreamConfig.ID,
+		c.upstreamConfig.HTTPURL,
+	).Set(float64(duration.Milliseconds()))
+
+	c.logger.Debug("Ran passive LatencyCheck.", zap.Any("upstreamID", c.upstreamConfig.ID), zap.Any("latency", duration), zap.Error(c.Err))
+}
+
+func (c *LatencyCheck) IsPassing() bool {
+	if c.errorCircuitBreaker.IsOpen() {
+		c.logger.Debug(
+			"LatencyCheck is not passing due to too many errors.",
+			zap.String("upstreamID", c.upstreamConfig.ID),
+			zap.Error(c.Err),
+		)
+
+		return false
+	}
+
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	// TODO(polsar): If one method's latency check is failing, the upstream will be marked as unhealthy,
+	//  which will affect all other methods. Is this what we want?
+	for method, breaker := range c.methodLatencyBreaker {
+		if breaker.IsOpen() {
+			c.logger.Debug(
+				"LatencyCheck is not passing due to high latency of an RPC method.",
+				zap.String("upstreamID", c.upstreamConfig.ID),
+				zap.Any("method", method),
+				zap.Error(c.Err),
+			)
+
+			return false
+		}
+	}
+
+	return true
+}
+
+func (c *LatencyCheck) RecordRequest(data *types.RequestData) {
+	if c.routingConfig.PassiveLatencyChecking {
+		return
+	}
+
+	latencyCircuitBreaker := c.getLatencyCircuitBreaker(data.Method)
+	latencyCircuitBreaker.RecordLatency(data.Latency)
+
+	if data.Latency >= latencyCircuitBreaker.GetThreshold() {
+		c.metricsContainer.LatencyCheckHighLatencies.WithLabelValues(
+			c.upstreamConfig.ID,
+			c.upstreamConfig.HTTPURL,
+			metrics.HTTPRequest,
+		).Inc()
+	}
+
+	var isError bool // false
+
+	if data.HTTPResponseCode >= http.StatusBadRequest {
+		// No RPC responses are available since the HTTP request errored out.
+		isError = c.isError(strconv.Itoa(data.HTTPResponseCode), "", "")
+	} else if data.ResponseBody != nil {
+		for _, resp := range data.ResponseBody.GetSubResponses() {
+			if resp.Error != nil {
+				// TODO(polsar): Should we ignore this response if it does not correspond to an RPC request?
```

Comment:
Ah gotcha yeah that's fine

Reply:
I was actually referring to this case where the response body can't be properly serialized to a JSON RPC object like [here ](https://github.com/satsuma-xyz/node-gateway/blob/51c61567667450e73e92bd3d6e7bf89e6d06f3aa/internal/route/request_executor.go#L241-L245) - ideally we should still check that response body if it contains errors strings. I've noticed that sometimes RPCs don't return a [properly formatted](https://github.com/satsuma-xyz/node-gateway/blob/b7f20aa2ad97f53772e9fa1565a300be7c0fff78/internal/jsonrpc/jsonrpc.go#L61-L66) JSON response when returning errors.

We can address this a separate PR as this one is getitng too big. As for priority, this is p important (v0/v1) because this [happens](https://alchemyinsights.slack.com/archives/C06F8P7P3FZ/p1724448080064109) often